modules/frontend/search_streaming.go

```diff
@@ -95,17 +96,74 @@ func (p *diffSearchProgress) finalResult() *shardedSearchResults {
 	return p.progress.result()
 }
 
+type multiProgress struct {
+	mu       sync.Mutex
+	progress []*atomic.Pointer[*diffSearchProgress]
+}
+
+func newMultiProgress() *multiProgress {
+	return &multiProgress{
+		progress: make([]*atomic.Pointer[*diffSearchProgress], 0),
+	}
+}
+
+func (mp *multiProgress) Add(p *atomic.Pointer[*diffSearchProgress]) {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	mp.progress = append(mp.progress, p)
+}
+
+// Snapshot returns a safe copy of progress list. Use this function to get a copy of progress list
+// and use it to do further processing on the progress objects.
+func (mp *multiProgress) Snapshot() []*atomic.Pointer[*diffSearchProgress] {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	copyList := make([]*atomic.Pointer[*diffSearchProgress], len(mp.progress))
+	copy(copyList, mp.progress)
+	return copyList
+}
+
+func (mp *multiProgress) finalResults(ctx context.Context) *shardedSearchResults {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	if len(mp.progress) == 1 {
+		// single tenant query, only one progress object.
+		p := *mp.progress[0].Load()
+		return p.finalResult()
+	}
+
+	// multi-tenant query, merge result from all progress objects.
+	// note: newSearchProgress is used to combine results from multiple progress objects, don't use it for anything else.
+	comb := newSearchProgress(ctx, 0, 0, 0, 0)
+	for _, progress := range mp.progress {
+		// FIXME: add a null check here??
+		p := *progress.Load()
+		r := p.finalResult()
+		comb.addResponse(r.response)
+		if r.err != nil || r.statusCode != http.StatusOK {
+			comb.setError(r.err)
+			comb.setStatus(r.statusCode, r.statusMsg)
+		}
+	}
+	return comb.result()
+}
+
 // newSearchStreamingGRPCHandler returns a handler that streams results from the HTTP handler
 func newSearchStreamingGRPCHandler(cfg Config, o overrides.Interface, downstream http.RoundTripper, reader tempodb.Reader, searchCache *frontendCache, apiPrefix string, logger log.Logger) streamingSearchHandler {
 	searcher := streamingSearcher{
-		logger:        logger,
-		downstream:    downstream,
-		reader:        reader,
-		postSLOHook:   searchSLOPostHook(cfg.Search.SLO),
-		o:             o,
-		searchCache:   searchCache,
-		cfg:           &cfg,
-		preMiddleware: newMultiTenantUnsupportedMiddleware(cfg, logger),
+		logger:      logger,
+		downstream:  downstream,
+		reader:      reader,
+		postSLOHook: searchSLOPostHook(cfg.Search.SLO),
+		o:           o,
+		searchCache: searchCache,
+		cfg:         &cfg,
+		// pass NoOp combiner because we combine results ourselves, and don't use
+		// combined results from middleware
+		preMiddleware: newMultiTenantMiddleware(cfg, combiner.NewNoOp, logger),
```

Comment:
multitenant middleware waits for all tenants to have responses which it then combines and returns. 

how does this work with streaming?

Reply:
we keep reference to progress objects and stream results via that, and we don't care about the results that it combines. that's why I am using a NoOp combiner that just throws away the results.

blocking in waits for all tenants to finish, and when it returns, we signal our select case to send the final result.

we again use progress objects to build final result.

Reply:
ah, i see now. i am a bit concerned about orphaning goroutines in the multi tenant middleware, but it makes sense.

it's about as gorpy as the rest of this code. :)